### ISSUE TITLE
"Error: ENOENT: no such file or directory" on logrotated file

### ISSUE THREAD

**Issue by crazy-man** [DoubleCheck/safe-log-reader#11]
Hello,
I tried to run simple example from README.md on non-existing file:
var filePath = "/tmp/dummy";
var someDir = "/tmp";

Here's output:
watching for /tmp/dummy to appear
created bookmark dir: /tmp/.bookmark
watching /tmp
events.js:160
      throw er; // Unhandled 'error' event
      ^

Error: ENOENT: no such file or directory, open '/tmp/dummy'
    at Error (native)

What I expected, is that script would wait till the file is created( seems like it starts watching parent dir (/tmp), but crashes after a couple of seconds

I tried to find where exception is thrown, but no stacktrace is available.

I'm running on debian 8, with nodejs v6.2.2.

**Comment by msimerson** [comment]
Odd, that's exactly the use case [this test](https://github.com/DoubleCheck/safe-log-reader/blob/master/test/reader.js#L265-L305) covers. 

When I use the example from the README.md, I get this output and it just sits there waiting until I Control-C it:

````
$ node foo
watching for /tmp/dummy to appear
watching /tmp
^C
````

Please provide more information about your environment.

**Comment by crazy-man** [comment]
I took clean Debian GNU/Linux 8.2 (jessie) without nodejs installed(Virtualbox image).
Linux debian8 3.16.0-4-amd64 #1 SMP Debian 3.16.7-ckt11-1+deb8u4 (2015-09-19) x86_64 GNU/Linux
Installed nodejs using this link:
https://nodejs.org/en/download/package-manager/

$ curl -sL https://deb.nodesource.com/setup_6.x | sudo -E bash -
$ sudo apt-get install -y nodejs

$ node -v
v6.2.2

installed safe-log-reader using npm:
$ npm i safe-log-reader

ran the script, and got:
$ node foo.js 
watching for /tmp/dummy to appear
created bookmark dir: /tmp/.bookmark
watching /tmp
events.js:160
      throw er; // Unhandled 'error' event
      ^

Error: ENOENT: no such file or directory, open '/tmp/dummy'
    at Error (native)

I also ran unit tests(using mocha), they pass:
  growing file
      ✓ reads exactly 1 line appended after EOF (142ms)
    after file rotation
      ✓ reads lines appended to new file rotate.log (168ms)
      - reads lines appended to rotated file
    on non-existent file
      ✓ discovers and reads (134ms)
    unreadable file
      ✓ reads nothing (104ms)
      ✓ does not watch (101ms)

This happened for both root and regular users.

If I create file, it starts OK. Then I move it:

$ mv /tmp/dummy /tmp/dummy2

The output is:
end of /tmp/dummy
watching /tmp/dummy
watching /tmp
events.js:160
      throw er; // Unhandled 'error' event


Initially I got this error when I ran log-ship-elastic-postfix module on postfix /var/log/mail.log
There was no activity on postfix, and I called logrotate
$ logrotate --force /etc/logrotate.d/rsyslog
new mail.log is created after a while, during this period, safe-log-reader crashed.

**Comment by crazy-man** [comment]
Seems like I found the cause:
I ran it with DEBUG flag:
$ DEBUG=1 node foo.js
watching for /tmp/dummy to appear
	resolveAncestor: /tmp
watching /tmp
watcher saw rename on sh-thd-1022457023
....
and crash

So I went to Reader.prototype.watchEvent function(line 278 in index.js)
there is no validation that filename that we got from watcher is our log file.

Adding:
 if(path.basename(this.filePath) != filename){
    return false;
  }
helped, but when file is removed:
$ rm /tmp/dummy
watchChange function is called
after that createStream is called and function Bookmark.prototype.read returns at line 102:
fs.stat(logFilePath, function (err, stat) {
  if (err) return done(err);

That's our error.

**Comment by msimerson** [comment]
Thanks for the additional debugging info.

I see a few things here.

1. you're storing log files in /tmp, which is, umm, an unusual choice if one is the least bit concerned with the safety / integrity of their logs. That gives rise to...
2. the underlying Linux mechanisms for "watching" a file don't permit one to watch a non-existing file, so the alternative is to watch the enclosing directory and wait for the file to appear.

> there is no validation that filename that we got from watcher is our log file.

Oops:

````
filename: missing.log
filePath: /Users/matt/Documents/git/doublecheck/safe-log-reader/test/data/missing.log
basename: missing.log
...
filename: rotate.log.1
filePath: /Users/matt/Documents/git/doublecheck/safe-log-reader/test/data/rotate.log
basename: rotate.log
````

Your implementation assumes the filename of the event matches (note the extension on rotate.log). Putting that check in watchEvent will break log rotation watching.

I seems prudent that when watching a directory, we **should** be filtering for events matching the filename. It just has to be implemented such that it doesn't break the file rename detection. Try this:

```js
diff --git a/index.js b/index.js
index 9d3ae00..668969c 100644
--- a/index.js
+++ b/index.js
@@ -267,6 +267,9 @@ Reader.prototype.watch = function (fileOrDir) {
     if (err) return logger.error(err);
 
     logger.info('watching ' + existentPath);
+
+    slr.isWatchingParent = existentPath === fileOrDir ? false : true;
+
     slr.watcher = fs.watch(
       existentPath,
       slr.watchOpts,
@@ -277,6 +280,12 @@ Reader.prototype.watch = function (fileOrDir) {
 
 Reader.prototype.watchEvent = function (event, filename) {
   logger.debug('watcher saw ' + event + ' on ' + filename);
+
+  if (this.isWatchingParent) {
+    // make sure event filename matches
+    if (filename !== path.basename(this.filePath)) return;
+  }
+
   switch (event) {
     case 'change':
       this.watchChange(filename);
```

**Comment by crazy-man** [comment]
1. I used /tmp directory to reproduce the issue. In real world /var/log/mail.log is watched. But it doesn't change anything, since there are other logs that are rapidly modified(syslog, messages, etc)
2. Yes, watching directory makes sense.
But... OOPS :)
```
watching /var/log/mail.log
bookmark.save: line: 5
watcher saw rename on mail.log
	resolveAncestor: /var/log
watching /var/log
watcher saw change on syslog
opening for read: /var/log/mail.log
events.js:160
      throw er; // Unhandled 'error' event
      ^

Error: ENOENT: no such file or directory, open '/var/log/mail.log'
```
The cause is in `Reader.prototype.renameLinux` (line 336 in index.js):
```        
// watch parent dir for file to reappear
slr.watch(path.dirname(slr.filePath));
```
This causes `isWatchingParent `to be `true`.

I fixed it by calling `watch` on original file. 
`slr.watch(slr.filePath);`
`resolveAncestor` function finds the correct directory:
```
watcher saw rename on mail.log
	resolveAncestor: /var/log
watching /var/log
watcher saw change on syslog
watcher saw rename on messages
watcher saw change on messages
watcher saw change on messages
watcher saw rename on mail.info.5.gz
watcher saw rename on mail.warn.5.gz
watcher saw rename on mail.log.5.gz
watcher saw rename on messages.5.gz
watcher saw change on syslog
watcher saw rename on mail.log
{ dev: 2049,
  mode: 33184,
  nlink: 1,
....
```
There is still some weird behavior when I just delete the file:
`$ rm /var/log//mail.log`
This triggers `change` event (and not `rename` as I would expect):
```
watcher saw change on mail.log
opening for read: /var/log/mail.log
events.js:160
      throw er; // Unhandled 'error' event
```
But it's not that critical for me.

**Comment by msimerson** [comment]
@crazy-man, would you mind assembling the combination of changes into a PR?

**Comment by msimerson** [comment]
Fixed by #13
